File: 2021/day04/day4.py
LENGTH_OF_BOARD = 5


def read_file():
    with open("input.in", "r") as file:
        numbers = tuple(map(int, file.readline().strip().split(",")))
        file.readline()
        list_of_board = []
        board = []
        for line in file.readlines():
            line = line.strip()
            if line == "":
                list_of_board.append(board)
                board = []
            else:
                board.append(list(map(int, line.split())))
        list_of_board.append(board)
    return numbers, list_of_board


def check_win(shadow_board):
    for row in shadow_board:
        if sum(row) == LENGTH_OF_BOARD:
            return True
    transposed_shadow_board = list(map(list, zip(*shadow_board)))
    for row in transposed_shadow_board:
        if sum(row) == LENGTH_OF_BOARD:
            return True
    return False


# Returns index of board that have won. If no board has won, return -1
def check_all_win(shadow_boards):
    winning_boards = []
    number_of_boards = len(shadow_boards)
    for i in range(number_of_boards):
        shadow_board = shadow_boards[i]
        if check_win(shadow_board):
            winning_boards.append(i)
    return winning_boards


def create_shadow_boards(boards):
    number_of_boards = len(boards)
    return [[[0]*5 for _ in range(LENGTH_OF_BOARD)] for _ in range(number_of_boards)]


def mark_board(number, boards, shadow_boards):
    number_of_boards = len(boards)
    for i in range(number_of_boards):
        for j in range(LENGTH_OF_BOARD):
            for k in range(LENGTH_OF_BOARD):
                if number == boards[i][j][k]:
                    shadow_boards[i][j][k] = 1
    return shadow_boards


def find_winner(numbers, boards, shadow_boards):
    past_numbers = set()
    for number in numbers:
        past_numbers.add(number)
        shadow_boards = mark_board(number, boards, shadow_boards)
        list_of_winning_index = check_all_win(shadow_boards)
        if len(list_of_winning_index) == 0:
            continue
        winner_index = list_of_winning_index[0]
        return number, past_numbers, boards[winner_index]
    return None, None, None


def find_loser(numbers, boards, shadow_boards):
    last_winning_number_index = -1
    last_winner = None
    for i in range(len(numbers)):
        number = numbers[i]
        shadow_boards = mark_board(number, boards, shadow_boards)
        list_of_winning_index = check_all_win(shadow_boards)
        if len(list_of_winning_index) != 0:
            last_winning_number_index = i
            last_winner = boards[list_of_winning_index[-1]]
        list_of_winning_index = list_of_winning_index[::-1]
        for winner_index in list_of_winning_index:
            del boards[winner_index]
            del shadow_boards[winner_index]
    # print(last_winner)
    # print(shadow_boards[winner_index])
    # print(last_winning_number_index, numbers[last_winning_number_index])
    # print(numbers[:last_winning_number_index])
    number = numbers[last_winning_number_index]
    past_numbers = numbers[:last_winning_number_index + 1]
    return number, past_numbers, last_winner


def get_result(number, past_numbers, winning_board):
    # Find sum of all numbers on board
    total_sum = sum([sum(row) for row in winning_board])
    flattened_board = [digit for row in winning_board for digit in row]
    sum_of_unmarked_numbers = total_sum
    for past_number in past_numbers:
        if past_number in flattened_board:
            sum_of_unmarked_numbers -= past_number
    return sum_of_unmarked_numbers * number


def print_boards(boards):
    for i in range(len(boards)):
        board = boards[i]
        print("Index: " + str(i))
        for row in board:
            print(row)
        print("\n")


def part1():
    numbers, boards = read_file()
    shadow_boards = create_shadow_boards(boards)
    number, past_numbers, winning_board = find_winner(numbers, boards, shadow_boards)
    print(f"Part 1 Answer: {get_result(number, past_numbers, winning_board)}")


def part2():
    numbers, boards = read_file()
    shadow_boards = create_shadow_boards(boards)
    number, past_numbers, losing_board = find_loser(numbers, boards, shadow_boards)
    print(f"Part 2 Answer: {get_result(number, past_numbers, losing_board)}")


if __name__ == "__main__":
    part1()
    part2()

File: 2021/day04/day4.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <sstream>
#include <tuple>
#include <numeric>

using namespace std;


int LENGTH_OF_BOARD = 5;

void part1(vector<int> numbers, vector<vector<vector<int>>> boards);
void part2(vector<int> numbers, vector<vector<vector<int>>> boards);
pair<vector<int>, vector<vector<vector<int>>>> readInput();
bool checkWin(vector<vector<vector<int>>> shadowBoards);
vector<int> checkAllWin(vector<vector<vector<int>>> shadowBoards);
vector<vector<vector<int>>> createShadowBoards(vector<vector<vector<int>>> boards);
vector<vector<vector<int>>> markBoard(int number, vector<vector<vector<int>>> boards, vector<vector<vector<int>>> shadowBoards);
tuple<int, vector<int>, vector<vector<int>>> findWinner(vector<int> numbers, vector<vector<vector<int>>> boards, vector<vector<vector<int>>> shadowBoards);
tuple<int, vector<int>, vector<vector<int>>> findLoser(vector<int> numbers, vector<vector<vector<int>>> boards, vector<vector<vector<int>>> shadowBoards);
int getResult(int number, vector<int> pastNumbers, vector<vector<int>> winningBoard);
vector<int> flattenBoard(const vector<vector<int>>& board);
void printBoards(const vector<vector<vector<int>>>& boards);


int main() {
    pair<vector<int>, vector<vector<vector<int>>>> rawInput = readInput();
    vector<int> numbers = rawInput.first;
    vector<vector<vector<int>>> boards = rawInput.second;
    part1(numbers, boards);
    part2(numbers, boards);
    return 0;
}

void part1(vector<int> numbers, vector<vector<vector<int>>> boards) {
    auto shadowBoards = createShadowBoards(boards);
    auto rawResult = findWinner(numbers, boards, shadowBoards);
    int result = getResult(get<0>(rawResult), get<1>(rawResult), get<2>(rawResult));
    cout << "Part 1: " << result << endl;
}

void part2(vector<int> numbers, vector<vector<vector<int>>> boards) {
    auto shadowBoards = createShadowBoards(boards);
    auto rawResult = findLoser(numbers, boards, shadowBoards);
    int result = getResult(get<0>(rawResult), get<1>(rawResult), get<2>(rawResult));
    cout << "Part 2: " << result << endl;
}

pair<vector<int>, vector<vector<vector<int>>>> readInput() {
    string rawNumbers; cin >> rawNumbers;
    vector<int> numbers;
    stringstream ss(rawNumbers);

    for (int i; ss >> i;) {
        numbers.push_back(i);
        if (ss.peek() == ',') {
            ss.ignore();
        }
    }

    vector<vector<vector<int>>> boards;
    vector<vector<int>> board;
    vector<int> row;
    int digit;
    int count = 1;
    while (cin >> digit) {
        row.push_back(digit);
        if (count % 5 == 0 && count != 0) {
            vector<int> rowToPush = row;
            board.push_back(rowToPush);
            row.clear();
        }
        if (count % 25 == 0 && count != 0) {
            vector<vector<int>> boardToPush = board;
            boards.push_back(boardToPush);
            board.clear();
        }
        count++;
    }

    pair<vector<int>, vector<vector<vector<int>>>> output;
    output.first = numbers;
    output.second = boards;
    return output;
}

bool checkWin(vector<vector<int>> shadowBoard) {
    // Check rows
    for (auto row : shadowBoard) {
        int rowSum = accumulate(row.begin(), row.end(), 0);
        if (rowSum == LENGTH_OF_BOARD) {
            return true;
        }
    }
    // Check columns
    for (int i = 0; i < LENGTH_OF_BOARD; i++) {
        int colSum = 0;
        for (int j = 0; j < LENGTH_OF_BOARD; j++) {
            colSum += shadowBoard.at(j).at(i);
        }
        if (colSum == LENGTH_OF_BOARD) {
            return true;
        }
    }
    return false;
}

vector<int> checkAllWin(vector<vector<vector<int>>> shadowBoards) {
    vector<int> winningBoards;
    int numberOfBoards = shadowBoards.size();
    for (int i = 0; i < numberOfBoards; i++) {
        vector<vector<int>> shadowBoard = shadowBoards.at(i);
        if (checkWin(shadowBoard)) {
            winningBoards.push_back(i);
        }
    }
    return winningBoards;
}

vector<vector<vector<int>>> createShadowBoards(vector<vector<vector<int>>> boards) {
    vector<vector<vector<int>>> shadowBoards;
    int numberOfBoards = boards.size();
    for (int i = 0; i < numberOfBoards; i++) {
        vector<vector<int>> board = {{0, 0, 0, 0, 0},
                                     {0, 0, 0, 0, 0},
                                     {0, 0, 0, 0, 0},
                                     {0, 0, 0, 0, 0},
                                     {0, 0, 0, 0, 0}};
        shadowBoards.push_back(board);
    }
    return shadowBoards;
}

vector<vector<vector<int>>> markBoard(int number, vector<vector<vector<int>>> boards, vector<vector<vector<int>>> shadowBoards) {
    int numberOfBoards = boards.size();
    for (int i = 0; i < numberOfBoards; i++) {
        for (int j = 0; j < LENGTH_OF_BOARD; j++) {
            for (int k = 0; k < LENGTH_OF_BOARD; k++) {
                if (number == boards[i][j][k]) {
                    shadowBoards[i][j][k] = 1;
                }
            }
        }
    }
    return shadowBoards;
}

tuple<int, vector<int>, vector<vector<int>>> findWinner(vector<int> numbers, vector<vector<vector<int>>> boards, vector<vector<vector<int>>> shadowBoards) {
    vector<int> pastNumbers;
    for (int number : numbers) {
        pastNumbers.push_back(number);
        vector<vector<vector<int>>> updatedShadowBoards = markBoard(number, boards, shadowBoards);
        shadowBoards = updatedShadowBoards;
        vector<int> winningIndices = checkAllWin(shadowBoards);
        if (winningIndices.empty()) {
            continue;
        }
        int winningIndex = winningIndices[0];
        return make_tuple(number, pastNumbers, boards[winningIndex]);
    }
    return make_tuple(-1, pastNumbers, boards[0]);
}

tuple<int, vector<int>, vector<vector<int>>> findLoser(vector<int> numbers, vector<vector<vector<int>>> boards, vector<vector<vector<int>>> shadowBoards) {
    int lastWinningIndex = -1;
    vector<vector<int>> lastWinner;
    for (int i = 0; i < numbers.size(); i++) {
        int number = numbers[i];
        auto updatedShadowBoards = markBoard(number, boards, shadowBoards);
        shadowBoards = updatedShadowBoards;
        vector<int> winningIndices = checkAllWin(shadowBoards);
        if (!winningIndices.empty()) {
            lastWinningIndex = i;
            lastWinner = boards[winningIndices[winningIndices.size() - 1]];
        }
        reverse(winningIndices.begin(), winningIndices.end());
        for (auto winningIndex : winningIndices) {
            boards.erase(boards.begin() + winningIndex);
            shadowBoards.erase(shadowBoards.begin() + winningIndex);
        }
    }
    int number = numbers[lastWinningIndex];
    vector<int> pastNumbers(numbers.begin(), numbers.begin() + lastWinningIndex + 1);
    return make_tuple(number, pastNumbers, lastWinner);
}

int getResult(int number, vector<int> pastNumbers, vector<vector<int>> winningBoard) {
    int totalSum = 0;
    for (auto row : winningBoard) {
        totalSum += accumulate(row.begin(), row.end(), 0);
    }
    vector<int> flattenedBoard = flattenBoard(winningBoard);
    int sumOfUnmarkedNumbers = totalSum;
    for (int pastNumber : pastNumbers) {
        if (find(flattenedBoard.begin(), flattenedBoard.end(), pastNumber) != flattenedBoard.end()) {
            sumOfUnmarkedNumbers -= pastNumber;
        }
    }
    return sumOfUnmarkedNumbers * number;
}

vector<int> flattenBoard(const vector<vector<int>>& board) {
    vector<int> flattenedBoard;
    for (auto row : board) {
        flattenedBoard.insert(flattenedBoard.end(), row.begin(), row.end());
    }
    return flattenedBoard;
}

void printBoards(const vector<vector<vector<int>>>& boards) {
    for (const auto& board: boards) {
        for (const auto& row: board) {
            for (auto digit: row) {
                cout << digit << " ";
            }
            cout << endl;
        }
        cout << endl;
    }
}
